README
For the python version you will need pyutil [1]. 

To build the C version you will need libzstr [2], a C library for safe, efficient,
8-bit clean strings.

Please see the DESIGN file for a justification of the specific base-32 encoding 
and why it differs from previous encodings such as [3].

[1] http://sourceforge.net/projects/pyutil
[2] http://sourceforge.net/projects/libzstr
[3] http://www.ietf.org/internet-drafts/draft-josefsson-base-encoding-04.txt


debian/changelog
python-zbase32 (1.1.23) lucid; urgency=low

  [ Ian Eure ]
  * Architecture: any; this has C extensions.

  [ SimpleGeo Nerds ]

 -- SimpleGeo Nerds <nerds@simplegeo.com>  Tue, 22 Jun 2010 16:24:02 +0000

python-zbase32 (1.1.22) lucid; urgency=low

  [ Ian Eure ]
  * Transplant build-versioning from SG.
  * Fold Makefile into GNUmakefile.
  * Fix build number in version file.
  * Add current version

  [ SimpleGeo Nerds ]

 -- SimpleGeo Nerds <nerds@simplegeo.com>  Tue, 22 Jun 2010 00:45:12 +0000

python-zbase32 (1.1.4) unstable; urgency=low

  * Nuke .egg-info directory.

 -- Ian Eure <ian@simplegeo.com>  Mon, 14 Jun 2010 13:13:45 -0700

python-zbase32 (1.1.3) unstable; urgency=low

  * Ignore build artifacts.

 -- Ian Eure <ian@simplegeo.com>  Mon, 14 Jun 2010 12:42:14 -0700

python-zbase32 (1.1.2) unstable; urgency=low

  * Remove Debian clean rule.

 -- Ian Eure <ian@simplegeo.com>  Mon, 14 Jun 2010 12:36:43 -0700

python-zbase32 (1.1.1-3) unstable; urgency=low

  * Build-depend on python-setuptools.

 -- Ian Eure <ian@simplegeo.com>  Fri, 11 Jun 2010 16:37:41 -0700

python-zbase32 (1.1.1-2) unstable; urgency=low

  * Remove SOURCES.txt, it doesn't belong here.

 -- Ian Eure <ian@simplegeo.com>  Mon, 17 May 2010 11:09:09 -0700

python-zbase32 (1.1.1-1) unstable; urgency=low

  * Debianized.

 -- Ian Eure <ian@simplegeo.com>  Mon, 17 May 2010 11:01:12 -0700


test.c
#include "base32.h"
#include "zstr.h"
#include "zutil.h"

#undef NDEBUG
#include <assert.h>
#include <ctype.h>
#include <math.h>
#include <stdlib.h>
#include <stdio.h>


zstr randz(const size_t len)
{
	zstr result = new_z(len);
	size_t i;
	for (i=0; i<len; i++) {
		result.buf[i] = rand() % 256;
	}
	return result;
}

int test_rands()
{
	int i;
	czstr z = cz(randz(16));
	/*czstr z = cs_as_cz("8b8t");*/
	czstr a;
	czstr zrecovered;
	for (i=0; i<32768; i++) {
		a = cz(b2a(z));
		/* printf("a.len: %d, buf: %s\n", a.len, a.buf);*/
		assert (strlen(cz_as_cs(a)) == divceil(z.len*8, 5));
		assert (isgraph(a.buf[0]));
		zrecovered = cz(a2b(a));
		if (!zeq(z, zrecovered)) {
			printf("failed basic recovery test. z: %s, a: %s, zr: %s\n", repr(z).buf, a.buf, repr(zrecovered).buf);
			return -1;
		}
		free((void*)a.buf);
		free((void*)zrecovered.buf);
	}
	return 0;
}

int test_b2alen_doesnt_overflow()
{
	if (b2alen(SIZE_T_MAX) != (unsigned long)ceil((double)SIZE_T_MAX/(double)5))
		return -1;
	else
		return 0;
}

int bench_ed()
{
	unsigned i;
	for (i=0;i<32;i++)
		test_rands();
	return 0;
}

int main(int argc, char**argv)
{
	if (test_b2alen_doesnt_overflow()) {
		return -1;
	}
	return test_rands();
	/*return bench_ed();*/
}


zbase32/test/test_zbase32.py
#!/usr/bin/env python
#
# Copyright (c) 2002-2005 Zooko "Zooko" Wilcox-O'Hearn
# mailto:zooko@zooko.com
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this work to deal in this work without restriction (including the rights
# to use, modify, distribute, sublicense, and/or sell copies).

import random, unittest

# http://zooko.com/repos/pyutil
from pyutil import randutil

from zbase32 import *
from zbase32 import b2a

def insecurerandstr(n):
    return ''.join(map(chr, map(random.randrange, [0]*n, [256]*n)))

class base32TestCase(unittest.TestCase):
    def test_ende(self):
        bs = insecurerandstr(2**3)
        cs=b2a(bs)
        bs2=a2b(cs)
        assert bs2 == bs, "bs2: %s, bs: %s" % (`bs2`, `bs`,)

    def test_ende_long(self):
        bs = insecurerandstr(2**3)
        cs=b2a_long(bs)
        bs2=a2b_long(cs)
        assert bs2 == bs, "bs2: %s, bs: %s" % (`bs2`, `bs`,)

    def test_both(self):
        bs = insecurerandstr(2**3)
        cs=b2a(bs)
        asl=b2a_long(bs)
        assert cs == asl, "cs: %s, asl: %s" % (`cs`, `asl`,)
        cs=b2a(bs)
        bs2=a2b(cs)
        bs2l=a2b_long(cs)
        assert bs2 == bs2l
        assert bs2 == bs

    def test_big(self):
        bs = insecurerandstr(2**9)
        cs=b2a(bs)
        asl=b2a_long(bs)
        assert cs == asl, "cs: %s, asl: %s" % (`cs`, `asl`,)
        cs=b2a(bs)
        bs2=a2b(cs)
        bs2l=a2b_long(cs)
        assert bs2 == bs2l
        assert bs2 == bs

    def DISABLED_test_odd_sizes_violates_preconditions(self):
        """
        The Python implementation of b2a_l() actually works if you pass any lengthinbits, and if you don't zero out the unused least significant bits.  The precondition assertions are (a) because the high speed C implementation does not work in those cases and (b) because it might help people discover errors in their own code.  Anyway, if you have assertion-checking turned off, then this test should pass, too.
        """
        for j in range(2**6):
            lib = random.randrange(1, 2**8)
            bs = insecurerandstr(random.randrange(1, 2**5))
            cs = b2a_l(bs, lib)
            assert len(cs) == (lib+4)/5 # the size of the base-32 encoding must be just right
            asl = b2a_l_long(bs, lib)
            assert len(asl) == (lib+4)/5 # the size of the base-32 encoding must be just right
            assert cs == asl
            bs2 = a2b_l(cs, lib)
            assert len(bs2) == (lib+7)/8 # the size of the result must be just right
            bs2l = a2b_l_long(cs, lib)
            assert len(bs2l) == (lib+7)/8 # the size of the result must be just right
            assert bs2 == bs2l
            assert trimnpad(bs, lib) == bs2, "trimnpad(%s, %s): %s, bs2: %s" % (`bs`, lib, `trimnpad(bs, lib)`, `bs2`,)

    def test_odd_sizes(self):
        for j in range(2**6):
            lib = random.randrange(1, 2**8)
            bs = insecurerandstr((lib+7)/8)
            # zero-out unused least-sig bits
            if lib%8:
                b=ord(bs[-1])
                b = b >> (8 - (lib%8))
                b = b << (8 - (lib%8))
                bs = bs[:-1] + chr(b)
            cs = b2a_l(bs, lib)
            assert len(cs) == (lib+4)/5 # the size of the base-32 encoding must be just right
            asl = b2a_l_long(bs, lib)
            assert len(asl) == (lib+4)/5 # the size of the base-32 encoding must be just right
            assert cs == asl
            bs2 = a2b_l(cs, lib)
            assert len(bs2) == (lib+7)/8 # the size of the result must be just right
            bs2l = a2b_l_long(cs, lib)
            assert len(bs2l) == (lib+7)/8 # the size of the result must be just right
            assert bs2 == bs2l
            assert trimnpad(bs, lib) == bs2, "trimnpad(%s, %s): %s, bs2: %s" % (`bs`, lib, `trimnpad(bs, lib)`, `bs2`,)

    def test_n8mo4_is_an_encoding(self):
        self.failUnless(could_be_base32_encoded_l('n8mo4', 25))

    def test_could_be(self):
        # base-32 encoded strings could be
        for j in range(2**9):
            rands = insecurerandstr(random.randrange(1, 2**7))
            randsenc = b2a(rands)
            assert could_be_base32_encoded(randsenc), "rands: %s, randsenc: %s, a2b(randsenc): %s" % (`rands`, `randsenc`, `a2b(randsenc)`,)

        # base-32 encoded strings with unusual bit lengths could be, too
        for j in range(2**9):
            bitl = random.randrange(1, 2**7)
            bs = insecurerandstr((bitl+7)/8)
            # zero-out unused least-sig bits
            if bitl%8:
                b=ord(bs[-1])
                b = b >> (8 - (bitl%8))
                b = b << (8 - (bitl%8))
                bs = bs[:-1] + chr(b)
            assert could_be_base32_encoded_l(b2a_l(bs, bitl), bitl)

        # anything with a bogus character couldn't be
        s = b2a(insecurerandstr(random.randrange(3, 2**7)))
        assert not could_be_base32_encoded('\x00' + s)
        assert not could_be_base32_encoded(s + '\x00')
        assert not could_be_base32_encoded(s[:1] + '\x00' + s[1:])
        assert not could_be_base32_encoded(s[:1] + '\x00' + s[2:])

        # a base-32 encoded string with an alleged lengthinbits of 16 but with 1-bits in the 18th, 19th,
        # and 20th location couldn't be.
        assert not could_be_base32_encoded_l('yyz', 16)

def _help_bench_e(N):
    return b2a(insecurerandstr(N))
def _help_bench_ed(N):
    return a2b(b2a(insecurerandstr(N)))
def _help_bench_e_l(N):
    return b2a_long(insecurerandstr(N))
def _help_bench_ed_l(N):
    return a2b_long(b2a_long(insecurerandstr(N)))

def benchem():
    import benchfunc # from pyutil
    print "e: "
    benchfunc.bench(_help_bench_e, TOPXP=13)
    print "ed: "
    benchfunc.bench(_help_bench_ed, TOPXP=13)
    print "e_l: "
    benchfunc.bench(_help_bench_e_l, TOPXP=13)
    print "ed_l: "
    benchfunc.bench(_help_bench_ed_l, TOPXP=13)

def suite():
    suite = unittest.makeSuite(base32TestCase, 'test')
    return suite

if __name__ == "__main__":
    unittest.main()



base32.h
/**
 * copyright 2002-2007 Zooko "Zooko" Wilcox-O'Hearn
 * mailto:zooko@zooko.com
 *
 * Permission is hereby granted, free of charge, to any person
 * obtaining a copy of this work to deal in this work without
 * restriction (including the rights to use, modify, distribute,
 * sublicense, and/or sell copies).
 */
#ifndef __INCL_base32_h
#define __INCL_base32_h

static int const base32_vermaj = 0;
static int const base32_vermin = 9;
static int const base32_vermicro = 14;
static char const* const base32_vernum = "0.9.14";

#include "zstr.h" /* http://zooko.com/repos/libzstr */
#include "zutil.h" /* http://zooko.com/repos/libzutil */

#include <assert.h>
#include <stddef.h>

/**
 * @param lengthinbits the length of the data in bits
 *
 * @return the length of the base-32 encoding of the data in characters
 *
 * @precondition lengthinbits must not be greater than
 */
size_t const b2alen(const size_t lengthinbits);
size_t const B2ALEN(const size_t lengthinbits);
#ifdef NDEBUG
#define b2alen(lengthinbits) ldivceil(lengthinbits, 5)
#define B2ALEN(lengthinbits) (LDIVCEIL((lengthinbits), 5))
#endif /* #ifdef NDEBUG */

/**
 * @param os the data to be encoded in a zstr
 *
 * @return the contents of `os' in base-32 encoded form in a (newly allocated) zstr
 */
zstr b2a(const czstr os);
#ifndef NDEBUG
zstr B2A(const czstr os);
#else
#define B2A(x) b2a_l((x), (x).len*8)
#endif /* #ifndef NDEBUG */

/**
 * @param cs the base-32 encoded data (a string)
 *
 * @return the binary data that was encoded into `cs' (a string)
 */
zstr a2b(const czstr cs);
#ifndef NDEBUG
zstr A2B(const czstr cs);
#else
#define A2B(x) a2b_l((x), (((x).len*5+3)/8)*8)
#endif /* #ifndef NDEBUG */

/**
 * @param os the data to be encoded in a zstr
 * @param lengthinbits the number of bits of data in `os' to be encoded
 *
 * b2a_l() will generate a base-32 encoded string big enough to encode lengthinbits bits.  So for 
 * example if os is 2 bytes long and lengthinbits is 15, then b2a_l() will generate a 3-character-
 * long base-32 encoded string (since 3 quintets is sufficient to encode 15 bits).  If os is 2 bytes
 * long and lengthinbits is 16 (or None), then b2a_l() will generate a 4-character string.  Note 
 * that `b2a_l()' does not mask off unused least-significant bits, so for example if os is 2 bytes 
 * long and lengthinbits is 15, then you must ensure that the unused least-significant bit of os is 
 * a zero bit or you will get the wrong result.  This precondition is tested by assertions if 
 * assertions are enabled.
 *
 * Warning: if you generate a base-32 encoded string with `b2a_l()', and then someone else tries to 
 * decode it by calling `a2b()' instead of  `a2b_l()', then they will (probably) get a different 
 * string than the one you encoded!  So only use `b2a_l()' when you are sure that the encoding and 
 * decoding sides know exactly which `lengthinbits' to use.  If you do not have a way for the 
 * encoder and the decoder to agree upon the lengthinbits, then it is best to use `b2a()' and 
 * `a2b()'.  The only drawback to using `b2a()' over `b2a_l()' is that when you have a number of 
 * bits to encode that is not a multiple of 8, `b2a()' can sometimes generate a base-32 encoded 
 * string that is one or two characters longer than necessary.
 *
 * @return the contents of `os' in base-32 encoded form in a (newly allocated) zstr
 *
 * On memory exhaustion, if Z_EXHAUST_EXIT is not defined, return a "null" zstr with .buf == NULL 
 * and .len == 0.
 */
zstr b2a_l(const czstr cs, const size_t lengthinbits);
zstr b2a_l_extra_Duffy(const czstr cs, const size_t lengthinbits);

/**
 *
 * @param lengthinbits the number of bits of data in encoded into `cs'
 *
 * a2b_l() will return a result big enough to hold lengthinbits bits.  So for example if cs is 
 * 4 characters long (encoding at least 15 and up to 20 bits) and lengthinbits is 16, then a2b_l() 
 * will return a string of length 2 (since 2 bytes is sufficient to store 16 bits).  If cs is 
 * 4 characters long and lengthinbits is 20, then a2b_l() will return a string of length 3 (since 
 * 3 bytes is sufficient to store 20 bits).  Note that `b2a_l()' does not mask off unused least- 
 * significant bits, so for example if cs is 4 characters long and lengthinbits is 17, then you must
 *  ensure that all three of the unused least-significant bits of cs are zero bits or you will get 
 * the wrong result.  This precondition is tested by assertions if assertions are enabled.  
 * (Generally you just require the encoder to ensure this consistency property between the least 
 * significant zero bits and value of `lengthinbits', and reject strings that have a length-in-bits 
 * which isn't a multiple of 8 and yet don't have trailing zero bits, as improperly encoded.)
 *
 * Please see the warning in the documentation of `b2a_l()' regarding the use of `b2a()' versus 
 * `b2a_l()'.
 *
 * @return a newly allocated zstr containing the data encoded in `cs'
 *
 * @precondition `cs' must be possibly base32 encoded data.: could_be_base32_encoded_l(cs, lengthinbits)
 *
 * On memory exhaustion, if Z_EXHAUST_EXIT is not defined, return a "null" zstr with .buf == NULL 
 * and .len == 0.
 */
zstr a2b_l(const czstr cs, const size_t lengthinbits);
zstr a2b_l_very_Duffy(const czstr cs, const size_t lengthinbits);

#endif /* #ifndef __INCL_base32_h */



zbase32/zbase32id.py
#  Copyright (c) 2002, 2003 Zooko "Zooko" Wilcox-O'Hearn
#  portions Copyright (c) 2001 Autonomous Zone Industries
#  This file is licensed under the
#    GNU Lesser General Public License v2.1.
#    See the file COPYING or visit http://www.gnu.org/ for details.

import string

from pyutil import humanreadable

import zbase32 

printableascii = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_=+!@#$%^&*()`~[]\{}|;':\",./<>? \t" # I just typed this in by looking at my keyboard.  It probably doesn't matter much if I missed some, because I only use it to guess whether a 20-byte string should be represented as a string or as an ID.  If all of the characters in the string are found `printableascii', then we guess that it is a string, not an id.
nulltrans = string.maketrans('', '')

def abbrev_bare(id, b2a_l=zbase32.b2a_l, trimnpad=zbase32.trimnpad):
    return b2a_l(trimnpad(id[:4], 25), 25)

def abbrev(id):
    return '<' + abbrev_bare(id) + '>'

def to_base32(id, b2a_l=zbase32.b2a_l):
    return b2a_l(id, 160)

def could_be_id(id, could_be_base32_encoded_l=zbase32.could_be_base32_encoded_l):
    return could_be_base32_encoded_l(id, 160)

def could_be_abbrev(id, could_be_base32_encoded_l=zbase32.could_be_base32_encoded_l):
    return could_be_base32_encoded_l(id, 25)

class AbbrevRepr(humanreadable.BetterRepr):
    def __init__(self):
        humanreadable.BetterRepr.__init__(self)

    def repr_str(self, obj, level, could_be_base32_encoded_l=zbase32.could_be_base32_encoded_l, nulltrans=nulltrans, printableascii=printableascii, abbrev=abbrev):
        if len(obj) == 20:
            # But maybe it was just a 20-character human-readable string, like "credit limit reached", so this is an attempt to detect that case.
            if len(obj.translate(nulltrans, printableascii)) == 0:
                if self.maxstring >= 22:
                    return `obj`

                # inlining repr.repr_string() here...
                s = `obj[:self.maxstring]`
                if len(s) > self.maxstring:
                    i = max(0, (self.maxstring-3)/2)
                    j = max(0, self.maxstring-3-i)
                    s = `obj[:i] + obj[len(obj)-j:]`
                    s = s[:i] + '...' + s[len(s)-j:]
                    # ... done inlining `repr.repr_string()'
                return s

            return abbrev(obj)
        elif could_be_base32_encoded_l(obj, 160):
            # new "z-base-32" base-32 encoding
            return '<' + obj[:5] + '>'
        else:
            # inlining repr.repr_string() here...
            s = `obj[:self.maxstring]`
            if len(s) > self.maxstring:
                i = max(0, (self.maxstring-3)/2)
                j = max(0, self.maxstring-3-i)
                s = `obj[:i] + obj[len(obj)-j:]`
                s = s[:i] + '...' + s[len(s)-j:]
            # ... done inlining `repr.repr_string()'
            return s

# use base32id repr to pretty-print ids embedded in structures
humanreadable.brepr = AbbrevRepr()



base32.c
/**
 * copyright 2002-2007 Zooko "Zooko" Wilcox-O'Hearn
 * mailto:zooko@zooko.com
 *
 * Permission is hereby granted to any person obtaining a copy of this work to
 * deal in this work without restriction (including the rights to use, modify,
 * distribute, sublicense, and/or sell copies).
 */
#include "base32.h"

#include <math.h>
#include <string.h>
#include <stdio.h> /* XXX only for debug printfs */

static const char*const chars="ybndrfg8ejkmcpqxot1uwisza345h769";
/* revchars: index into this table with the ASCII value of the char.  The result is the value of that quintet. */
static const zbyte revchars[]={ 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 18, 255, 25, 26, 27, 30, 29, 7, 31, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 24, 1, 12, 3, 8, 5, 6, 28, 21, 9, 10, 255, 11, 2, 16, 13, 14, 4, 22, 17, 19, 255, 20, 15, 0, 23, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, };

#undef b2alen
size_t const
b2alen(const size_t lengthinbits) {
	return divceil(lengthinbits, 5);
}

#undef B2ALEN
size_t const
B2ALEN(const size_t lengthinbits) {
	return divceil(lengthinbits, 5);
}

zstr
b2a(const czstr os) {
	return b2a_l(os, os.len*8);
}

#undef B2A
zstr
B2A(const czstr os) {
	return b2a_l(os, os.len*8);
}

zstr
b2a_l(const czstr os, const size_t lengthinbits) {
		return b2a_l_extra_Duffy(os, lengthinbits);
}

zstr
b2a_l_extra_Duffy(const czstr os, const size_t lengthinbits) {
	zstr result = new_z(DIVCEIL(os.len*8, 5)); /* if lengthinbits is not a multiple of 8 then this is allocating space for 0, 1, or 2 extra quintets that will be truncated at the end of this function if they are not needed */
#ifndef Z_EXHAUST_EXIT
	if (result.buf == NULL)
		return result;
#endif /* #ifndef Z_EXHAUST_EXIT */

	zbyte* resp = result.buf + result.len; /* pointer into the result buffer, initially pointing to the "one-past-the-end" quintet */
	const zbyte* osp = os.buf + os.len; /* pointer into the os buffer, initially pointing to the "one-past-the-end" octet */

	/* Now this is a real live Duff's device.  You gotta love it. */
	unsigned long x=0; /* to hold up to 32 bits worth of the input */
	switch ((osp - os.buf) % 5) {
	case 0:
		do {
			x = *--osp;
			*--resp = chars[x % 32]; /* The least sig 5 bits go into the final quintet. */
			x /= 32; /* ... now we have 3 bits worth in x... */
	case 4:
			x |= ((unsigned long)(*--osp)) << 3; /* ... now we have 11 bits worth in x... */
			*--resp = chars[x % 32];
			x /= 32; /* ... now we have 6 bits worth in x... */
			*--resp = chars[x % 32];
			x /= 32; /* ... now we have 1 bits worth in x... */
	case 3:
			x |= ((unsigned long)(*--osp)) << 1; /* The 8 bits from the 2-indexed octet.  So now we have 9 bits worth in x... */
			*--resp = chars[x % 32];
			x /= 32; /* ... now we have 4 bits worth in x... */
	case 2:
			x |= ((unsigned long)(*--osp)) << 4; /* The 8 bits from the 1-indexed octet.  So now we have 12 bits worth in x... */
			*--resp = chars[x%32];
			x /= 32; /* ... now we have 7 bits worth in x... */
			*--resp = chars[x%32];
			x /= 32; /* ... now we have 2 bits worth in x... */
	case 1:
			x |= ((unsigned long)(*--osp)) << 2; /* The 8 bits from the 0-indexed octet.  So now we have 10 bits worth in x... */
			*--resp = chars[x%32];
			x /= 32; /* ... now we have 5 bits worth in x... */
			*--resp = chars[x];
		} while (osp > os.buf);
	} /* switch ((osp - os.buf) % 5) */

	/* truncate any unused trailing zero quintets */
	result.len = DIVCEIL(lengthinbits, 5);
	result.buf[result.len] = '\0';
	return result;
}

zstr
a2b(const czstr cs) {
	return a2b_l(cs, ((cs.len*5+3)/8)*8);
}

#undef A2B
zstr
A2B(const czstr cs) {
	return a2b_l(cs, ((cs.len*5+3)/8)*8);
}

zstr
a2b_l(const czstr cs, const size_t lengthinbits) {
	return a2b_l_very_Duffy(cs, lengthinbits);
}

zstr
a2b_l_very_Duffy(const czstr cs, const size_t lengthinbits) {
	unsigned long x=0; /* to hold up to 32 bits worth of the input */
	zstr result = new_z(divceil(cs.len*5, 8)); /* if lengthinbits is not a multiple of 5 then this is allocating space for 0 or 1 extra octets that will be truncated at the end of this function if they are not needed */
#ifndef Z_EXHAUST_EXIT
	if (result.buf == NULL)
		return result;
#endif /* #ifndef Z_EXHAUST_EXIT */

	zbyte* resp = result.buf + result.len; /* pointer into the result buffer, initially pointing to the "one-past-the-end" octet */
	const zbyte* csp = cs.buf + cs.len; /* pointer into the cs buffer, initially pointing to the "one-past-the-end" character */

	fflush(stdout);

	/* Now this is a real live Duff's device.  You gotta love it. */
	switch ((csp - cs.buf) % 8) {
	case 0:
		fflush(stdout);
		do {
			fflush(stdout);
			x = revchars[*--csp]; /* 5 bits... */
	case 7:
			fflush(stdout);
			x |= revchars[*--csp] << 5; /* 10 bits... */
			*--resp = x % 256;
			x /= 256; /* 2 bits... */
	case 6:
			fflush(stdout);
			x |= revchars[*--csp] << 2; /* 7 bits... */
	case 5:
			fflush(stdout);
			x |= revchars[*--csp] << 7; /* 12 bits... */
			*--resp = x % 256;
			x /= 256; /* 4 bits... */
	case 4:
			x |= revchars[*--csp] << 4; /* 9 bits... */
			*--resp = x % 256;
			fflush(stdout);
			x /= 256; /* 1 bit... */
	case 3:
			x |= revchars[*--csp] << 1; /* 6 bits... */
			fflush(stdout);
	case 2:
			x |= revchars[*--csp] << 6; /* 11 bits... */
			*--resp = x % 256;
			fflush(stdout);
			x /= 256; /* 3 bits... */
	case 1:
			x |= revchars[*--csp] << 3; /* 8 bits... */
			*--resp = x % 256;
			fflush(stdout);
		} while (csp > cs.buf);
	} /* switch ((csp - cs.buf) % 8) */

	/* truncate any unused trailing zero octets */
	result.len = DIVCEIL(lengthinbits, 8);
	result.buf[result.len] = '\0';
	return result;
}

